src/backend/metal/src/command.rs

```diff
@@ -736,13 +742,22 @@ impl RawCommandQueue<Backend> for CommandQueue {
             None
         };
 
-        let buffers = submit.cmd_buffers.into_iter().collect::<Vec<_>>();
-        let num_buffers = buffers.len();
-        for (i, buffer) in buffers.into_iter().enumerate() {
-            let buffer = buffer.borrow();
-            let command_buffer: &metal::CommandBufferRef = match buffer.inner_ref().sink {
-                 CommandSink::Immediate { ref cmd_buffer, .. } => cmd_buffer,
-                 CommandSink::Deferred { ref passes, .. } => {
+        for buffer in submit.cmd_buffers {
+            let inner = buffer.borrow().inner();
+            let command_buffer: &metal::CommandBufferRef = match inner.sink {
+                CommandSink::Immediate { ref cmd_buffer, .. } => {
+                    // schedule the retained buffers to release after the commands are done
+                    if !inner.retained_buffers.is_empty() {
+                        let retained_buffers = mem::replace(&mut inner.retained_buffers, Vec::new());
+                        let release_block = ConcreteBlock::new(move |_cb: *mut ()| -> () {
+                            let _ = retained_buffers; // move and auto-release
+                        }).copy();
+                        let cb_ref: &metal::CommandBufferRef = cmd_buffer;
+                        msg_send![cb_ref, addCompletedHandler: release_block.deref() as *const _];
+                    }
+                    cmd_buffer
+                }
+                CommandSink::Deferred { ref passes, .. } => {
```

Comment:
Doesn't `update_buffer` still cause `retained_buffers` to become populated if the application is currently recording into `CommandSink::Deferred`? Or is this avoided somehow?

Reply:
Correct, but for multi-use buffer (which is what `Deferred` is for) we don't want to set up the completion handler - retained buffer vectors is going to be cleaned up on `reset()`

Reply:
Right, got it, thanks. Updated commit looks good to me then!